README.md
# Free Champion Rotations for League of Legends

### Installation docs can be found in docs/installation.md, please read them!

## About the project

* Displays the free champions available for the current week, for normal games and ARAM
* Displays Aram rotations (65 champions + current week. Duplicates if they exist)
* Has an option for both EUW and NA
* Includes a countdown timer for both NA and EUW
* Has a toggle for dark/light mode (caches if light mode is chosen)
* Detects geolocation if no cookie exist (puts you on either NA or EU)
* images has link to their op.gg guides
* no ads
* The app do not collect any information- one cookie by default (geolocate on what server to put you, and remembers),
other cookie is only created if light mode is chosen.
* Can also be run with docker-compose

## How it will look like
![Example of site layout](https://i.imgur.com/XGC8WJA.png)

## Important changes - 31/03/2020
- Riot changed how ARAM rotations works as of 31/03/2020, it is available under **legacy** branch
- major difference between legacy and current system is that now it displays 65 champions
that are always available in ARAM, plus the weekly rotation from before.
Which makes the previous two rotations that used to be cached obsolete.



CHANGELOG.md
# Changelog


---

## All notable changes to this publication will be documented in this file

### 1.8.8 - 11/03/2021

* Added Roboto from google fonts as a default font-family was missing

### 1.8.7 - 01/08/2020

* Added an example.png in the README.md showing how the layout will look like when its on a website

### 1.8.7 - 18/07/2020

* Swapped php-fpm image with an Alpine image, changed user/permission to make container more secure.

### 1.8.6 - 12/07/2020

* removed some actions from the Dockerfile (for the php-fpm/app), including cron- as it isn't set up for the Docker stack i've made (just for development/testing).

### 1.8.5 - 11/07/2020

* Added and fixed errors with permissions in new docker containers, changed php-fpm image to dockerfile build for permission args

### 1.8.4 - 10/07/2020

* Added ob_start to prevent problems with modified header information in docker container. Also added new dockerfiles that is isolated, new images seperately for webserver (nginx), php-fpm and composer.

### 1.8.3 - 02/06/2020

* Added alt text attribute to champion images for better access

### 1.8.2 - 13/05/2020

* Change in dockerfile to put specific ownership for the Cache dir to www-data, problem with file_put_contents and file_get_contents because ownership was set to 1000 in this dir

### 1.8.1 - 08/05/2020

* Fixed css img div issue where the container would be too big because of champion name length, adjusted with rem to fit the viewport

### 1.8 - 05/05/2020

* Optimized the docker container issues. Had issue with crashing containers because of CMD/Entrypoint in the images used, put all in one and added an infinite bash loop to keep the web container alive. Refactored so its built of an ubuntu image for stability and easier accessibility. It installs nginx, php-fpm through APT, and composer is downloaded and installed through curl

### 1.8b - 04/05/2020

* Fixed an issue with docker containers crashing making it only accessible on windows, but not on unix-like system because of permissions(chown/chmod), added a kamikaze container that will crash itself after its intended purpose of setting the environment correctly

### 1.8a -  03/05/2020

* Added testing environment for docker-compose, creates two containers, installs composer dependencies also.
* Fixed error with chown/chmod for docker container not being able to write, group and user set to www-data with 775.
* Added a new dockerfile based off the composer image

### 1.7.6 - 26/04/2020

* Added form validation with htmlspecialchar for POST action

### 1.7.5 - 20/04/2020

* Added a twig templating for errors that can occur if using VPN (server error 500) \ 
needs further investigation, some IPs will cause the geolocation API to not run (internal 500)

### 1.7.4 - 09/04/2020

* Added bin/ directory with two executable php scripts made to be used with crontab, see readme
* Added docs/ directory for installation.md

### 1.7.3 - 08/04/2020

* Refactored some code, added new dir with resources for Cache and other files that are not PHP

### 1.7.2 - 07/04/2020

* Fixed Readme with some vital information and made time window for rotation to 5min
* Added backup methods for fetching rotation despite being of opposite server location, 
would not work before because no cookie or post request sent from wget (with crontab), see readme for example

### 1.7.1 - 07/04/2020

* Fixed issue with overwriting current json rotation
* Changed writing from fread/fwrite/fclose to file_put_contents


### 1.7 - 06/04/2020

* Added an if statement that checks if the caches exist, and loads then despite the timer logic if they dont
as they are required to display anything in the first place, also added TODOS.
* Added for devs an auto-fetcher for external IP with phpdotenv if localhost is used to prevent geolocation to break
* Added a checker for if cookies exist, if not- methods for getIP and fetch(location) will invoke and put user at their
respective location, this cookie checker was added to prevent the geolocation API to be ratelimited

### 1.6 - 04/04/2020

* Refactored code so API usage to only load in once a week to prevent rate limiting
* Still needs testing to see how behaviour is under rotation, branching to **experimental**

### 1.5.2 - 04/04/2020

* Updated Twig template to display actual errors, tested with guzzle mocking

### 1.5.1 -  03/04/2020

* Fixed typo in guzzleclass making champions display the ID from DDragon JSON instead of the name,
the ID is still intended for the op.gg img link - where name would not work.

### 1.5 - 02/04/2020

* BIG CHANGE: Riot changed the format for ARAM, new is 65 set champions + current week from API. I refactored the
aramchampions method to display these 65 (removing duplicates). This will also make the cache not fully needed, only
for displaying the currentweek when the time difference between EU and NA happens.

### 1.4.3 - 01/04/2020

* Made a script that automatically fetches the most recent available DDragon patch and puts it in the phpdotenv

### 1.4.2 - 01/04/2020

* Refactored class code, removed constants and replaced with env variables for current patch

### 1.4.1 - 31/03/2020

* Added php built-in function for setting timezone to Europe/Berlin, DST should follow accordingly
* Added comments with what variables that should be changed if no DST

### 1.4 - 30/03/2020

* Added templates for Twig in case of Client/Server errors, foundation for future development- make template for the most common API errors possible

### 1.3.2 - 23/03/2020

* Fixed an error that displayed the wrong time for EUW and NA on the servercheck, correctly fixed to W - 1 day -2h for EUW and additional -6h for NA

### 1.3.1 - 22/03/2020

* Added a guzzle middleware for Riotgames REST API rate limit, max 3 request per second
* Added href on images to op.gg to the respective champion
* Added img for github with href to project in top-left corner

### 1.3.0 - 21/03/2020


src/Classes/Guzzleclass.php
<?php

namespace App\Classes;

use DateTime;
use GuzzleHttp\Client;
use Psr\Http\Message\ResponseInterface;
use Psr\Http\Message\StreamInterface;
use GuzzleHttp\Exception\ClientException;
use GuzzleHttp\Exception\ServerException;
use Twig\Environment;
use Twig\Loader\FilesystemLoader;

class Guzzleclass
{

    private $id;
    private $content;
    private $img;

    /**
     * @var Environment
     */
    private Environment $twig;
    /**
     * @var array|false|string
     */
    private $key;
    /**
     * @var ResponseInterface
     */
    private ResponseInterface $responseNA;
    /**
     * @var ResponseInterface
     */
    private ResponseInterface $responseEUW;
    /**
     * @var StreamInterface
     */
    private StreamInterface $euwBody;
    /**
     * @var StreamInterface
     */
    private StreamInterface $naBody;


    public function testAPI()
    {
        /**
         * check if guzzle can fetch riot API with ENV key
         * then if both NA and EUW has connection without errors
         */

        //API key loading
        $this->key = getenv('API'); // API KEY from .env
        if ($this->key == "API_KEY_HERE" || $this->key == null) {
            die("<span class=\"httpError\">Please put your actual API key in the .env file.</span>");
        }

        //Twig
        $loader = new FilesystemLoader('../templates');
        $this->twig = new Environment($loader);

        //if cache does not exist at all, fetch them
        $cacheEUW = dirname(__FILE__) . "/../../resources/Cache/rotationEUW.json";
        $cacheNA = dirname(__FILE__) . "/../../resources/Cache/rotationNA.json";
        if (!file_exists($cacheEUW)) {
            $this->requestEUW();
        } elseif (!file_exists($cacheNA)) {
            $this->requestNA();
        }

        // Doublechecks REST API in case serverlocation location is the opposite (only EU/NA available)
        // works without doing post request/having cookies existing- mainly for production server
        $this->remoteNA();
        $this->remoteEUW();

        //loading timers for 4min to change json with REST API
        $this->euTimer();
        $this->naTimer();
    }

    private function euTimer()
    {
        $currentTime = new DateTime();
        $startTime = new DateTime('Tue 01:58');
        $endTime = new DateTime('Tue 02:03');

        if (
            $currentTime->format('D H:i:s') >= $startTime->format('D H:i:s')
            && $currentTime->format('D H:i:s') <= $endTime->format('D H:i:s')
        ) {
            $this->requestEUW();
        }
    }

    private function natimer()
    {
        $currentTime = new DateTime();
        $startTime = new DateTime('Tue 10:58');
        $endTime = new DateTime('Tue 11:03');

        if (
            $currentTime->format('D H:i:s') >= $startTime->format('D H:i:s')
            && $currentTime->format('D H:i:s') <= $endTime->format('D H:i:s')
        ) {
            $this->requestNA();
        }
    }

    private function requestNA()
    {

        try {
            $client = new Client();
            $this->responseNA = $client->request(
                'GET',
                'https://na1.api.riotgames.com/lol/platform/v3/champion-rotations?api_key=' . $this->key
            );
        } catch (ClientException | ServerException $e) {
            $code = $e->getResponse()->getStatusCode();

            if (
                $code == 400 || $code == 401 || $code == 403 ||
                $code == 404 || $code == 405 || $code == 415 || $code == 429
            ) {
                echo $this->twig->render(
                    'error.html.twig',
                    [
                        'error' => 'Client',
                        'code' => $code,
                        'server' => 'NA'
                    ]
                );
            } elseif ($code == 500 || $code == 502 || $code == 503 || $code == 504) {
                echo $this->twig->render(
                    'error.html.twig',
                    [
                        'error' => 'Server',
                        'code' => $code,
                        'server' => 'NA'
                    ]
                );
            }
        }
        if ($this->responseNA->getStatusCode() == 200) {
            $this->naBody = $this->responseNA->getBody();
        }
        $file = dirname(__FILE__) . "/../../resources/Cache/rotationNA.json";
        file_put_contents($file, $this->naBody);
    }

    private function requestEUW()
    {

        try {
            $client = new Client();
            $this->responseEUW = $client->request(
                'GET',
                'https://euw1.api.riotgames.com/lol/platform/v3/champion-rotations?api_key=' . $this->key
            );
        } catch (ClientException | ServerException $e) {
            $code = $e->getResponse()->getStatusCode();

            if (
                $code == 400 || $code == 401 || $code == 403 ||
                $code == 404 || $code == 405 || $code == 415 || $code == 429
            ) {
                echo $this->twig->render(
                    'error.html.twig',
                    [
                        'error' => 'Client',
                        'code' => $code,
                        'server' => 'EUW'
                    ]
                );
            } elseif ($code == 500 || $code == 502 || $code == 503 || $code == 504) {
                echo $this->twig->render(
                    'error.html.twig',
                    [
                        'error' => 'Server',
                        'code' => $code,
                        'server' => 'EUW'
                    ]
                );
            }
        }
        if ($this->responseEUW->getStatusCode() == 200) {
            $this->euwBody = $this->responseEUW->getBody();
        }
        $file = dirname(__FILE__) . "/../../resources/Cache/rotationEUW.json";
        file_put_contents($file, $this->euwBody);
    }

    private function remoteNA()
    {
            $currentTime = new DateTime();
            $startTime = new DateTime('Tue 10:58');
            $endTime = new DateTime('Tue 11:53');

        if (
                $currentTime->format('D H:i:s') >= $startTime->format('D H:i:s')
                && $currentTime->format('D H:i:s') <= $endTime->format('D H:i:s')
        ) {
            $cacheNA = dirname(__FILE__) . "/../../resources/Cache/rotationNA.json";
            $array = json_decode(file_get_contents($cacheNA), true);
            $this->id = $array['freeChampionIds'];

            // Data Dragon JSON
            $champions = "http://ddragon.leagueoflegends.com/cdn/" . getenv('PATCH') . "/data/en_US/champion.json";
            $ddragon = new Client();
            $res = $ddragon->get($champions);
            $ddragon_json = $res->getBody();
            $this->content = json_decode($ddragon_json, true);
        }
    }

    private function remoteEUW()
    {
        $currentTime = new DateTime();
        $startTime = new DateTime('Tue 01:58');
        $endTime = new DateTime('Tue 02:03');

        if (
            $currentTime->format('D H:i:s') >= $startTime->format('D H:i:s')
            && $currentTime->format('D H:i:s') <= $endTime->format('D H:i:s')
        ) {
            $cacheEUW = dirname(__FILE__) . "/../../resources/Cache/rotationEUW.json";
            $array = json_decode(file_get_contents($cacheEUW), true);
            $this->id = $array['freeChampionIds'];

            // Data Dragon JSON
            $champions = "http://ddragon.leagueoflegends.com/cdn/" . getenv('PATCH') . "/data/en_US/champion.json";
            $ddragon = new Client();
            $res = $ddragon->get($champions);
            $ddragon_json = $res->getBody();
            $this->content = json_decode($ddragon_json, true);
        }
    }



    public function fetchID()
    {
        if (isset($_POST['EUW']) || isset($_COOKIE['EUW'])) {
            $cacheEUW = dirname(__FILE__) . "/../../resources/Cache/rotationEUW.json";
            $array = json_decode(file_get_contents($cacheEUW), true);
            $this->id = $array['freeChampionIds'];
        } elseif (isset($_POST['NA']) || isset($_COOKIE['NA'])) {
            $cacheNA = dirname(__FILE__) . "/../../resources/Cache/rotationNA.json";
            $array = json_decode(file_get_contents($cacheNA), true);
            $this->id = $array['freeChampionIds'];
        }

        // Data Dragon JSON
        $champions = "http://ddragon.leagueoflegends.com/cdn/" . getenv('PATCH') . "/data/en_US/champion.json";
        $ddragon = new Client();
        $res = $ddragon->get($champions);
        $ddragon_json = $res->getBody();
        $this->content = json_decode($ddragon_json, true);
    }


    public function currentWeek()
    {
        $this->img = "https://ddragon.leagueoflegends.com/cdn/" . getenv('PATCH') . "/img/champion/";
        foreach ($this->content['data'] as $champ) {
            foreach ($this->id as $freeid) {
                if ($champ["key"] == $freeid) {
                    $displayImg = $this->img . $champ["id"] . ".png"; ?>
                    <div class="item">
                        <?php
                        print_r("<a href=https://euw.op.gg/champion/{$champ['id']}/statistics/ target=_blank>
                        <img alt={$champ['id']} src={$displayImg}></a>");
                        ?>
                        <span class="caption"><?php echo $champ['name']; ?></span>
                    </div><?php
                }
            }
        }
    }


    public function aramChampions()
    {
        $json = dirname(__FILE__) . "/../../resources/alwaysaram.json";
            $json_array = json_decode(file_get_contents($json), true);
            $result = $json_array['always'];

        // removing duplicates, checking for diff between the ones fetched from the API and the ones merged
        $aramonly = array_merge($this->id, $result);
        $aram = array_unique($aramonly);
        $finalresult = array_diff($aram, $this->id);

        //prints champion name/img
        foreach ($this->content['data'] as $champ) {
            foreach ($finalresult as $freeid) {
                if ($champ["key"] == $freeid) {
                    $displayImg = $this->img . $champ["id"] . ".png"; ?>
                    <div class="item">
                    <?php
                    print_r("<a href=https://euw.op.gg/champion/{$champ['id']}/statistics/ target=_blank>
                        <img alt={$champ['id']} src={$displayImg}></a>");
                    ?>
                    <span class="caption"><?php echo $champ['name']; ?></span>
                    </div><?php
                }
            }
        }
    }

    public function serverCheck()
    {
        if (isset($_POST['EUW']) || isset($_COOKIE["EUW"])) {
            echo 'EUW';
        } elseif (isset($_POST['NA']) || isset($_COOKIE['NA'])) {
            echo 'NA';
        }
    }
}